test/e2e/framework/util.go

```diff
@@ -1284,3 +1285,110 @@ func taintExists(taints []v1.Taint, taintToFind *v1.Taint) bool {
+	ExpectNoError(err, "Failed to fetch initial resource")
+	listWatcher := &cache.ListWatch{
+		WatchFunc: func(listOptions metav1.ListOptions) (watch.Interface, error) {
+			return dc.Resource(resourceType).Namespace(namespace).Watch(ctx, listOptions)
+		},
+	}
+	resourceWatch, err := watchtools.NewRetryWatcher(initResource.GetResourceVersion(), listWatcher)
+	ExpectNoError(err, "Failed to create a resource watch of %v in namespace %v", resourceType.Resource, namespace)
+
+	// NOTE value of 3 retries seems to make sense
+	retries := 3
+retriesLoop:
+	for try := 1; try <= retries; try++ {
+		// NOTE the test may need access to the events to see what's going on, such as a change in status
+		actualWatchEvents := scenario(resourceWatch)
+		errs := sets.NewString()
+		// watchEventsLoop:
+		totalValidWatchEvents := 0
+		actualWatchEventsHasDelete := false
+		for watchEventIndex := range expectedWatchEvents {
+			foundExpectedWatchEvent := false
+			ExpectEqual(len(expectedWatchEvents) <= len(actualWatchEvents), true, "Error: actual watch events amount (%d) must be greater than or equal to expected watch events amount (%d)", len(actualWatchEvents), len(expectedWatchEvents))
+			for actualWatchEventIndex := range actualWatchEvents {
+				if actualWatchEvents[watchEventIndex].Type == expectedWatchEvents[actualWatchEventIndex].Type {
+					foundExpectedWatchEvent = true
+				}
+				if actualWatchEvents[actualWatchEventIndex].Type == watch.Deleted && actualWatchEventsHasDelete != true {
+					actualWatchEventsHasDelete = true
+				}
+			}
+			if foundExpectedWatchEvent == false {
+				errs.Insert(fmt.Sprintf("Watch event %v not found", expectedWatchEvents[watchEventIndex].Type))
+			}
+			totalValidWatchEvents++
+		}
+		err := retryCleanup()
+		ExpectNoError(err, "Error occurred when cleaning up resources")
+		if errs.Len() > 0 && try < retries {
+			fmt.Println("invariants violated:\n", strings.Join(errs.List(), "\n - "))
+			continue retriesLoop
+		}
+		ExpectEqual(errs.Len() > 0, false, strings.Join(errs.List(), "\n - "))
+		ExpectEqual(totalValidWatchEvents, len(expectedWatchEvents), "Error: there must be an equal amount of total valid watch events (%d) and expected watch events (%d)", totalValidWatchEvents, len(expectedWatchEvents))
+		break retriesLoop
+	}
+}
+
+// WatchUntilWithoutRetry ...
+// reads items from the watch until each provided condition succeeds, and then returns the last watch
+// encountered. The first condition that returns an error terminates the watch (and the event is also returned).
+// If no event has been received, the returned event will be nil.
+// Conditions are satisfied sequentially so as to provide a useful primitive for higher level composition.
+// Waits until context deadline or until context is canceled.
+//
+// the same as watchtools.UntilWithoutRetry, just without the closing of the watch - as for the purpose of being paired with WatchEventSequenceVerifier, the watch is needed for continual watch event collection
```

Comment:
not sure replicating a method whose documentation ends with "Consider making this function private to prevent misuse when the other occurrences in our codebase are gone." into our test framework for reuse is the best idea... :)

Reply:
This function is meant to be used inside tests which use the watch tooling also provided in this PR. The function appears to work as expected in my cases from my testing, checkout [the usage here](https://github.com/kubernetes/kubernetes/pull/90880/files#diff-41baf4b031431ab2c17b13c0e32aa223R150-R157)